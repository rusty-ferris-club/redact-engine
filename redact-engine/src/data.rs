//! Common structs
use serde_derive::Deserialize;

/// Default redact placeholder
pub const REDACT_PLACEHOLDER: &str = "[TEXT_REDACTED]";

#[derive(Debug, Deserialize, Clone)]
/// Describe redaction by Pattern
pub struct Pattern {
    #[serde(with = "serde_regex")]
    /// regex Pattern
    pub test: regex::Regex,
    /// capture group to redact
    pub group: usize,
}

#[derive(Debug, Deserialize, Clone)]
/// Redact information
pub struct Info {
    /// redacted string
    pub string: String,
    /// captures information
    pub captures: Vec<Captures>,
}

#[derive(Debug, Deserialize, Clone)]
/// Capture details
pub struct Captures {
    /// the captured text
    pub text: String,
    /// Position capture details
    pub position: Option<Position>,
}

#[derive(Debug, Deserialize, Clone)]
pub struct Position {
    /// capture line number
    pub line: usize,
    /// start caption position
    pub start_offset: usize,
    /// end caption position
    pub end_offset: usize,
}
