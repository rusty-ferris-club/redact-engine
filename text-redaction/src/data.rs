use serde_derive::Deserialize;

/// Default redact placeholder
pub const REDACT_PLACEHOLDER: &str = "[TEXT_REDACTED]";

#[derive(Debug, Deserialize, Clone)]
pub struct Pattern {
    #[serde(with = "serde_regex")]
    pub test: regex::Regex,
    pub group: usize,
}

#[derive(Debug, Deserialize, Clone)]
pub struct Info {
    pub string: String,
    pub captures: Vec<Captures>,
}

#[derive(Debug, Deserialize, Clone)]
pub struct Captures {
    pub text: String,
    pub position: Option<Position>,
}

#[derive(Debug, Deserialize, Clone)]
pub struct Position {
    pub line_number: usize,
    pub start_offset: usize,
    pub end_offset: usize,
}
