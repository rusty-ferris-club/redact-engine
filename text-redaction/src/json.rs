//! Redact text from JSON format
//!
//! # Optional
//! This requires `serde_json` feature to be enabled.
use anyhow::Result;
use serde_json::Value;

use crate::data::REDACT_PLACEHOLDER;

pub struct Redact {
    /// redact placeholder text
    pub text_placeholder: String,
    /// list of keys to reduct
    pub keys: Vec<String>,
    /// list of JSON specific path
    pub path: Vec<String>,
    /// list of JSON prefix path
    pub path_prefix: Vec<String>,
}

impl Default for Redact {
    /// Create a [`Redact`] Methods
    fn default() -> Self {
        Self::new(REDACT_PLACEHOLDER, vec![], vec![], vec![])
    }
}

impl Redact {
    pub fn with_redact_placeholder(text_placeholder: &str) -> Self {
        Self::new(text_placeholder, vec![], vec![], vec![])
    }

    /// Create a [`Redact`] Methods with all available fields
    pub fn new(
        text_placeholder: &str,
        keys: Vec<String>,
        path: Vec<String>,
        path_prefix: Vec<String>,
    ) -> Self {
        Self {
            text_placeholder: text_placeholder.to_string(),
            keys,
            path,
            path_prefix,
        }
    }

    /// redact JSON value bt givenkeys
    ///
    /// # Examples
    ///
    /// ```rust
    /// # use text_redaction::Redaction;
    /// let text = "foo,bar";
    ///
    /// Redaction::new().add_keys(vec!["key"]);
    /// # ;
    /// ```
    pub fn add_keys(mut self, keys: Vec<&str>) -> Self {
        self.keys.extend(keys.iter().map(|&s| s.to_string()));
        self
    }

    /// redact JSON value by specific key path list
    ///
    /// # Example:
    /// ## Redact by specific key
    /// {
    ///     "a": {
    ///         "b": {
    ///             "key": "redact_me",
    ///         },
    ///         "foo": "bar",
    ///         "key": "skip-redaction"
    ///     },
    ///     "key": "skip-redaction"
    /// }
    /// ```rust
    /// # use text_redaction::{Redaction, Pattern};
    /// # use regex::Regex;
    ///
    /// Redaction::new().add_paths(vec!["a.b.key"]);
    /// # ;
    /// ```
    /// ## Redact all keys under `a`
    /// {
    ///     "a": {
    ///         "b": {
    ///             "key": "redact_me",
    ///         },
    ///         "foo": "bar",
    ///         "key": "skip-redaction"
    ///     },
    ///     "key": "skip-redaction"
    /// }
    /// ```rust
    /// # use text_redaction::{Redaction, Pattern};
    /// # use regex::Regex;
    ///
    /// Redaction::new().add_paths(vec!["a.*"]);
    /// # ;
    pub fn add_paths(mut self, path: Vec<&str>) -> Self {
        for path in path.iter() {
            if path.ends_with('*') {
                self.path_prefix.push((*path).to_string().replace(".*", ""));
            } else {
                self.path.push((*path).to_string());
            }
        }

        self
    }

    /// redact json str
    pub fn redact_str(&self, str: &str) -> Result<String> {
        let mut json_value: Value = serde_json::from_str(str)?;
        // let start_path = String::new();
        self.redact_value(&mut json_value, String::new());
        Ok(json_value.to_string())
    }

    /// redact Value values
    fn redact_value(&self, json: &mut Value, path: String) {
        if let Some(obj) = json.as_object_mut() {
            obj.iter_mut().for_each(|(key, value)| {
                let mut obj_path = path.clone();
                // create a json key path to be able redact by path keys.
                if obj_path.is_empty() {
                    obj_path.push_str(&key.to_string());
                } else {
                    obj_path.push_str(&format!(".{}", key));
                };

                if self.path.contains(&obj_path) || self.path_prefix.contains(&obj_path) {
                    *value = Value::String(self.text_placeholder.to_string());
                } else if self.keys.contains(key) {
                    if value.is_array() {
                        self.redact_value_array(value);
                    } else {
                        *value = Value::String(self.text_placeholder.to_string());
                    }
                } else if value.is_object() {
                    self.redact_value(value, obj_path.clone());
                }
            });
        }
    }

    /// redact all Value array values
    fn redact_value_array(&self, array: &mut Value) {
        array.as_array_mut().iter_mut().for_each(|values| {
            values.iter_mut().for_each(|val| {
                *val = Value::String(self.text_placeholder.to_string());
            });
        });
    }
}

#[cfg(test)]
mod test_redaction {

    use insta::assert_debug_snapshot;
    use serde_json::json;

    use super::*;

    #[test]
    fn can_redact_value_by_key() {
        let json = json!({
            "bar": "baz",
            "key": "value",
        })
        .to_string();

        let redact = Redact::default().add_keys(vec!["bar"]);

        assert_debug_snapshot!(redact.redact_str(&json));
    }

    #[test]
    fn can_redact_value_by_path() {
        let json = json!({
        "a": {
            "b": {
                "key": "redact_me",
            },
            "foo": "bar",
            "key": "skip-redaction"
        },
        "key": "skip-redaction"
            })
        .to_string();

        let redact = Redact::default().add_paths(vec!["a.foo"]);

        assert_debug_snapshot!(redact.redact_str(&json));
    }

    #[test]
    fn can_redact_value_by_prefix_path() {
        let json = json!({
        "a": {
            "b": {
                "key": "redact_me",
            },
            "foo": "bar",
            "key": "skip-redaction"
        },
        "key": "skip-redaction1"
            })
        .to_string();

        let redact = Redact::default().add_paths(vec!["a.*"]);

        assert_debug_snapshot!(redact.redact_str(&json));
    }

    #[test]
    fn can_redact_value_combination() {
        let json = json!({
        "foo": {
            "b": {
                "key": "redact_me",
            },
            "foo": "redact_me",
            "key": "redact_me",
        },
        "bar": {
            "b": {
                "key": "skip-redaction",
            },
            "foo": "skip-redaction",
            "key": "redact_me"
        },
        "key": "redact_me",
        "baz": "skip-redaction"
        })
        .to_string();

        let redact = Redact::default()
            .add_paths(vec!["foo.*", "bar.key"])
            .add_keys(vec!["key"]);

        assert_debug_snapshot!(redact.redact_str(&json));
    }

    #[test]
    fn can_redact_value_array() {
        let redact = Redact::default();

        let mut array_value = Value::Array(vec![
            serde_json::Value::String("value-1".to_string()),
            serde_json::Value::String("value-2".to_string()),
        ]);
        redact.redact_value_array(&mut array_value);
        assert_debug_snapshot!(array_value);
    }
}
