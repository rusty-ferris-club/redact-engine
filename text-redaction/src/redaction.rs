//! redaction function user interface
use std::{io, str};

use anyhow::{bail, Result};
use regex::{escape, Regex};

#[cfg(feature = "redact-info")]
use crate::data::Info;
#[cfg(feature = "redact-json")]
use crate::json;
use crate::{
    data::{Pattern, REDACT_PLACEHOLDER},
    pattern,
};

/// Define redact settings
pub struct Redaction {
    /// Define an option to redact text in JSON schema. enable by `redact-json`
    /// feature flag enabled.
    #[cfg(feature = "redact-json")]
    json: json::Redact,

    /// Define the default redact option by patterns logic.
    pattern: pattern::Redact,
}

impl Default for Redaction {
    /// Create a [`Redaction`] Methods
    ///
    /// # Example
    ///
    /// ```rust
    /// use text_redaction::Redaction;
    /// Redaction::default()
    /// # ;
    /// ```
    fn default() -> Self {
        Self::new()
    }
}

impl Redaction {
    #[must_use]
    /// Create a [`Redaction`] Methods
    ///
    /// # Example
    ///
    /// ```rust
    /// use text_redaction::Redaction;
    /// Redaction::custom("CUSTOM_HIDDEN_TEXT")
    /// # ;
    /// ```
    pub fn new() -> Self {
        Self::custom(REDACT_PLACEHOLDER)
    }

    #[must_use]
    /// Create a [`Redaction`] with redact placeholder text.
    ///
    /// # Arguments
    /// * `redact_placeholder` - placeholder redaction
    ///
    /// # Example
    ///
    /// ```rust
    /// use text_redaction::Redaction;
    /// Redaction::custom("[HIDDEN_VALUE]")
    /// # ;
    /// ```
    pub fn custom(redact_placeholder: &str) -> Self {
        Self {
            #[cfg(feature = "redact-json")]
            json: json::Redact::with_redact_placeholder(redact_placeholder),

            pattern: pattern::Redact::with_redact_placeholder(redact_placeholder),
        }
    }

    /// redact exact string match
    ///
    /// # Arguments
    /// * `value` - The redaction value
    ///
    /// # Example
    ///
    /// ```rust
    /// use text_redaction::Redaction;
    /// let text = "foo,bar";
    /// Redaction::new().add_value("foo");
    /// # ;
    /// ```
    /// # Errors
    /// when the value could not converted to a regex
    pub fn add_value(self, value: &str) -> Result<Self> {
        let pattern = Pattern {
            test: Regex::new(&format!("({})", escape(value)))?,
            group: 1,
        };

        Ok(self.add_pattern(pattern))
    }

    /// redact exact string match from list of strings
    ///
    /// # Arguments
    /// * `values` - List of redaction value
    ///
    /// # Example
    ///
    /// ```rust
    /// use text_redaction::Redaction;
    /// let text = "foo,bar,baz";
    /// Redaction::new().add_values(vec!["foo", "baz"]);
    /// # ;
    /// ```
    /// # Errors
    /// when the value could not converted to a regex
    pub fn add_values(self, values: Vec<&str>) -> Result<Self> {
        let mut errors = vec![];

        let patterns = values
            .iter()
            .filter_map(|val| match Regex::new(&format!("({})", escape(val))) {
                Ok(test) => Some(Pattern { test, group: 1 }),
                Err(_e) => {
                    errors.push((*val).to_string());
                    None
                }
            })
            .collect::<Vec<_>>();

        if !errors.is_empty() {
            bail!("could not parse {} to regex", errors.join(","))
        }

        Ok(self.add_patterns(patterns))
    }

    #[must_use]
    /// Add a [`Pattern`] to the redaction list
    ///
    /// # Arguments
    /// * `pattern` - redact [Pattern]
    ///
    /// # Example
    ///
    /// ```rust
    /// use text_redaction::{Redaction, Pattern};
    /// use regex::Regex;
    /// let text = "foo,bar";
    /// let pattern = Pattern {
    ///    test: Regex::new("(bar)").unwrap(),
    ///    group: 1,
    /// };
    ///
    /// Redaction::new().add_pattern(pattern);
    /// # ;
    /// ```
    pub fn add_pattern(mut self, pattern: Pattern) -> Self {
        self.pattern = self.pattern.add_pattern(pattern);
        self
    }

    #[must_use]
    /// Add list if [`Pattern`] to the redaction list
    ///
    /// # Arguments
    /// * `patterns` - List of redact [Pattern]
    ///
    /// # Example
    ///
    /// ```rust
    /// use text_redaction::{Redaction, Pattern};
    /// use regex::Regex;
    /// let text = "foo,bar";
    /// let pattern = Pattern {
    ///    test: Regex::new("(bar)").unwrap(),
    ///    group: 1,
    /// };
    ///
    /// Redaction::new().add_patterns(vec![pattern]);
    /// # ;
    /// ```
    pub fn add_patterns(mut self, patterns: Vec<Pattern>) -> Self {
        self.pattern = self.pattern.add_patterns(patterns);
        self
    }

    #[cfg(feature = "redact-json")]
    #[must_use]
    /// Redact the JSON value of the given key. enable by `redact-json`
    ///
    /// # Optional
    /// When `redact-json` feature flag is enabled
    ///
    /// # Arguments
    /// * `key` -  The JSON key
    ///
    /// # Example
    ///
    /// ```rust
    /// use text_redaction::Redaction;
    /// Redaction::new().add_key("bar").add_key("array");
    /// # ;
    /// ```
    pub fn add_key(mut self, key: &str) -> Self {
        self.json = self.json.add_key(key);
        self
    }

    #[cfg(feature = "redact-json")]
    #[must_use]
    /// Redact the JSON by JSON path. enable by `redact-json`.
    ///
    /// # Optional
    /// When `redact-json` feature flag is enabled
    ///
    /// # Example
    ///
    /// ```rust
    /// use text_redaction::Redaction;
    /// Redaction::new().add_key("bar").add_key("array");
    /// # ;
    /// ```
    pub fn add_path(mut self, key: &str) -> Self {
        self.json = self.json.add_path(key);
        self
    }

    #[must_use]
    /// Redact from string
    pub fn redact_str(&self, str: &str) -> String {
        self.pattern.redact_patterns(str, false).string
    }

    #[cfg(feature = "redact-info")]
    #[must_use]
    /// Redact from string with extra information of the matches
    ///
    /// # Optional
    /// When `redact-info` feature flag is enabled
    pub fn redact_str_with_info(&self, str: &str) -> Info {
        self.pattern.redact_patterns(str, true)
    }

    /// Redact text from reader
    ///
    /// # Errors
    /// - When file not exists.
    /// - Could not open reader.
    pub fn redact_reader<R>(&self, rdr: R) -> Result<String>
    where
        R: io::Read,
    {
        let mut rdr_box = Box::new(rdr);
        let mut buffer = Vec::new();
        rdr_box.read_to_end(&mut buffer)?;
        Ok(self.redact_str(str::from_utf8(&buffer)?))
    }

    /// Redact text from reader with extra information of the matches
    ///
    /// # Optional
    /// When `redact-info` feature flag is enabled
    ///
    /// # Errors
    /// - When file not exists.
    /// - Could not open reader.
    #[cfg(feature = "redact-info")]
    pub fn redact_reader_with_info<R>(&self, rdr: R) -> Result<Info>
    where
        R: io::Read,
    {
        let mut rdr_box = Box::new(rdr);
        let mut buffer = Vec::new();
        rdr_box.read_to_end(&mut buffer)?;
        Ok(self.redact_str_with_info(str::from_utf8(&buffer)?))
    }

    #[cfg(feature = "redact-json")]
    /// Redact from string.
    ///
    /// # Optional
    /// When `redact-json` feature flag is enabled
    ///
    /// # Errors
    /// return an error when the given str is not a JSON string
    pub fn redact_json(&self, str: &str) -> Result<String> {
        self.json.redact_str(&self.redact_str(str))
    }
}

#[cfg(test)]
mod test_redaction {

    use std::{env, fs::File, io::Write};

    use insta::assert_debug_snapshot;

    use super::*;

    const TEXT: &str = "foo,bar,baz,extra";

    #[cfg(feature = "redact-json")]
    use serde_json::json;

    #[test]
    fn test_by_pattern() {
        let pattern = Pattern {
            test: Regex::new("(foo)").unwrap(),
            group: 1,
        };
        let patterns = vec![
            Pattern {
                test: Regex::new("(bar)").unwrap(),
                group: 1,
            },
            Pattern {
                test: Regex::new("(baz)").unwrap(),
                group: 1,
            },
        ];
        assert_debug_snapshot!(Redaction::new()
            .add_pattern(pattern)
            .add_patterns(patterns)
            .redact_str(TEXT));
    }

    #[test]
    fn test_bt_value() {
        assert_debug_snapshot!(Redaction::new()
            .add_value("foo")
            .unwrap()
            .add_values(vec!["bar", "baz"])
            .unwrap()
            .redact_str(TEXT));
    }

    #[test]
    fn can_redact_str() {
        let pattern = Pattern {
            test: Regex::new("(bar)").unwrap(),
            group: 1,
        };
        let redaction = Redaction::new().add_pattern(pattern);
        assert_debug_snapshot!(redaction.redact_str(TEXT));
    }

    #[test]
    #[cfg(feature = "redact-info")]
    fn can_redact_str_with_info() {
        let pattern = Pattern {
            test: Regex::new("(bar)").unwrap(),
            group: 1,
        };
        let redaction = Redaction::new().add_pattern(pattern);
        assert_debug_snapshot!(redaction.redact_str_with_info(TEXT));
    }

    #[test]
    fn can_redact_reader() {
        let file_path = env::temp_dir().join("foo.txt");

        let mut f = File::create(&file_path).unwrap();
        #[allow(clippy::unused_io_amount)]
        f.write(TEXT.as_bytes()).unwrap();

        let pattern = Pattern {
            test: Regex::new("(bar)").unwrap(),
            group: 1,
        };

        let redaction = Redaction::new().add_pattern(pattern);
        assert_debug_snapshot!(redaction.redact_reader(File::open(file_path).unwrap()));
    }

    #[test]
    #[cfg(feature = "redact-info")]
    fn can_redact_reader_with_info() {
        let file_path = env::temp_dir().join("foo.txt");

        let mut f = File::create(&file_path).unwrap();
        #[allow(clippy::unused_io_amount)]
        f.write(TEXT.as_bytes()).unwrap();

        let pattern = Pattern {
            test: Regex::new("(bar)").unwrap(),
            group: 1,
        };

        let redaction = Redaction::new().add_pattern(pattern);
        assert_debug_snapshot!(redaction.redact_reader_with_info(File::open(file_path).unwrap()));
    }

    #[test]
    fn can_redact_with_multiple_patterns() {
        let patterns = vec![
            Pattern {
                test: Regex::new("(bar)").unwrap(),
                group: 1,
            },
            Pattern {
                test: Regex::new("(foo),(bar),(baz)").unwrap(),
                group: 3,
            },
        ];

        let redaction = Redaction::new().add_patterns(patterns);
        assert_debug_snapshot!(redaction.redact_str(TEXT));
    }

    #[test]
    fn can_redact_with_placeholder_text() {
        let pattern = Pattern {
            test: Regex::new("(bar)").unwrap(),
            group: 1,
        };
        let redaction = Redaction::custom("[HIDDEN_TEXT]").add_pattern(pattern);
        assert_debug_snapshot!(redaction.redact_str(TEXT));
    }

    #[test]
    #[cfg(feature = "redact-json")]
    fn can_redact_json() {
        let pattern = Pattern {
            test: Regex::new("(redact-by-pattern)").unwrap(),
            group: 1,
        };

        let json = json!({
        "all-path": {
            "b": {
                "key": "redact_me",
            },
            "foo": "redact_me",
            "key": "redact_me",
        },
        "specific-key": {
            "b": {
                "key": "skip-redaction",
            },
            "foo": "skip-redaction",
            "key": "redact_me"
        },
        "key": "redact_me",
        "skip": "skip-redaction",
        "by-value": "bar",
        "by-pattern": "redact-by-pattern",
        })
        .to_string();

        let redaction = Redaction::default()
            .add_pattern(pattern)
            .add_path("all-path.*")
            .add_path("specific-key.key")
            .add_key("key")
            .add_value("bar")
            .unwrap();
        assert_debug_snapshot!(redaction.redact_json(&json));
    }
}
