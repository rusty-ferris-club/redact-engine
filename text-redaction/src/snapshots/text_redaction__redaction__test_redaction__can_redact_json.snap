---
source: text-redaction/src/redaction.rs
expression: redaction.redact_json(&json)
---
Ok(
    "{\"all-path\":\"[TEXT_REDACTED]\",\"by-pattern\":\"[TEXT_REDACTED]\",\"by-value\":\"[TEXT_REDACTED]\",\"key\":\"[TEXT_REDACTED]\",\"skip\":\"skip-redaction\",\"specific-key\":{\"b\":{\"key\":\"[TEXT_REDACTED]\"},\"foo\":\"skip-redaction\",\"key\":\"[TEXT_REDACTED]\"}}",
)
